pkg/ccl/changefeedccl/alter_changefeed_stmt.go

```diff
@@ -305,15 +167,443 @@ func alterChangefeedPlanHook(
+			droppedTargetSpans, err := fetchSpansForDescs(ctx, p, opts, statementTime, droppedTargetDescs)
+			if err != nil {
+				return nil, nil, nil, err
+			}
+			removeSpansFromProgress(newJobProgress, droppedTargetSpans)
+		}
+	}
+
+	for _, targetDesc := range newTargets {
+		targetName := tree.NewUnresolvedName(targetDesc.GetName())
+		newTargetList.Tables = append(newTargetList.Tables, targetName)
+	}
+
+	if err := validateNewTargets(ctx, p, newTargetList, newJobProgress, newJobStatementTime); err != nil {
+		return nil, nil, nil, err
+	}
+
+	return newTargetList, &newJobProgress, &newJobStatementTime, nil
+}
+
+func validateNewTargets(
+	ctx context.Context,
+	p sql.PlanHookState,
+	newTargets *tree.TargetList,
+	jobProgress jobspb.Progress,
+	jobStatementTime hlc.Timestamp,
+) error {
+	if len(newTargets.Tables) == 0 {
+		return pgerror.New(pgcode.InvalidParameterValue, "cannot drop all targets")
+	}
+
+	// when we resume the changefeed, we need to ensure that the newly added
+	// targets can be resolved at the time of the high watermark. If the high
+	// watermark is empty, then we need to ensure that the newly added targets can
+	// be resolved at the StatementTime of the changefeed job.
+	var resolveTime hlc.Timestamp
+	highWater := jobProgress.GetHighWater()
+	if highWater != nil && !highWater.IsEmpty() {
+		resolveTime = *highWater
+	} else {
+		resolveTime = jobStatementTime
+	}
+
+	allDescs, err := backupresolver.LoadAllDescs(ctx, p.ExecCfg(), resolveTime)
+	if err != nil {
+		return errors.Wrap(err, `error while validating new targets`)
+	}
+	descResolver, err := backupresolver.NewDescriptorResolver(allDescs)
+	if err != nil {
+		return errors.Wrap(err, `error while validating new targets`)
+	}
+
+	for _, targetName := range newTargets.Tables {
+		_, found, err := getTargetDesc(ctx, p, descResolver, targetName)
+		if err != nil {
+			return errors.Wrap(err, `error while validating new targets`)
+		}
+		if !found {
+			if highWater != nil && !highWater.IsEmpty() {
+				return errors.Errorf(`target %q cannot be resolved as of the high water mark. Please wait until the high water mark progresses past the creation time of this target in order to add it to the changefeed.`, tree.ErrString(targetName))
+			}
+			return errors.Errorf(`target %q cannot be resolved as of the creation time of the changefeed. Please wait until the high water mark progresses past the creation time of this target in order to add it to the changefeed.`, tree.ErrString(targetName))
+		}
+	}
+
+	return nil
+}
+
+// generateNewProgress determines if the progress of a changefeed job needs to
+// be updated based on the targets that have been added, the options associated
+// with each target we are adding/removing (i.e. with initial_scan or
+// no_initial_scan), and the current status of the job. If the progress does not
+// need to be updated, we will simply return the previous progress and statement
+// time that is passed into the function
+func generateNewProgress(
```

Comment:
I think this function would flow better readability-wise if it was structured as:
```go
inInitialBackfill := prevHighWater == nil || prevHighWater.IsEmpty()
inSchemaChangeBackfill := !inInitialBackfill && ....

if withInitialBackfill {
  // No changes needed, the new spans will join the existing backfill
  if InBackfill {
    return prevProgress, prevStatementTime, nil
  }

  // ... comment explaining why this wouldn't work
  if inSchemaChangeBackfill {
    // return error
  }

  // If we'd like to perform an initial scan on new targets ....
  newStatementTime = *prevHighWater
  newProgress := ... 
  return newProgress, newStatementTime, nil
}


// If we don't want a scan of new targets but we're in an ongoing backfill,
// ensure new spans are skipped by adding it to the checkpoint
if inInitialBackfill || inSchemaChangeBackfill {
    var expandedCheckpoint roachpb.SpanGroup
    ... // initialize expanded checkpoint

    newProgress := ... 
    return newProgress, prevStatementTime, nil
}
  
// Without any backfills to consider, no changes are needed
return prevProgress, prevStatementTime, nil

```

Not that you'd need to do _exactly_ this, but friction I ran into reading the current flow that I wanted to reduce was:
- Reading later code, seeing `spanGroup`, going "wait what is that", then scrolling up to figure out that it meant "all the old checkpointed spans and new spans", and then having to read each block to confirm that "ok this is only used as an expanded checkpoint".  Having it be called `expandedCheckpoint` makes it easier to infer without scrolling to its definition, and having it be scoped to the context where its always used helps the reader to know that its not relevant information anywhere else.
- Reading the contents of the first if statement going "ok why does this make sense if the high water is empty -> oh ok its cause we're in a backfill", and the second going "ok this block is for an empty checkpoint, but the error says its a schema change, and ok looks like this block shoudl only run for a schema change since we've handled the initial backfill already".  The condition is using specific data to infer an environment (some backfill), while the contents of the block are operating under the assumption of that entire environment rather than just that specific data, so it helps to make the data->environment link explicit by moving them into their own variables.  It also helps highlight potential issues, such as "we're saying its a schema change backfill if theres a checkpoint outside of initial scan, but what if this runs during a schema change backfill but before a checkpoint is laid down?".  

---

And on that note, have we tested that case out?  Would we just have 2 backfills of the new targets, one as it approaches statement time and another once it sees the BACKFILL boundary event to trigger the schema backfill?  Would the checkpoint from the initial backfill on the old spans be cleared out so as to not interfere with the schema backfill?

Reply:
Thanks for the suggestion! I updated the function and I believe it should be easier to understand now. Also with the case that you mentioned in this comment, since we are adding targets with a non-empty high watermark and empty checkpoint, we will:
* reset the high watermark
* set the statement time to the previous high watermark
* add all existing target spans to the checkpoint

As a result, I believe that this case would be equivalent to the case in which we are performing an original initial scan and a schema change occurs during that initial scan.